### PR TITLE
Add an upstream_repository field to METADATA.toml

### PR DESCRIPTION
**Pull request by srittau**
Closes: #10478

### CODE REVIEW

File: tests/parse_metadata.py [hunk 1]
```diff
@@ -120,6 +120,7 @@ class StubMetadata:
     requires: Annotated[list[str], "The raw requirements as listed in METADATA.toml"]
     extra_description: str | None
     stub_distribution: Annotated[str, "The name under which the distribution is uploaded to PyPI"]
+    upstream_repository: str | None
```
Comment (AlexWaygood) [inline tests/parse_metadata.py:123]:
```suggestion
    upstream_repository: Annotated[str, "The URL of the upstream repository"] | None
```

**Review by AlexWaygood — APPROVED** [review]

**Comment by srittau** [comment]
Script to add the existing URLs:

```python
import asyncio
from pathlib import Path

import aiohttp

from scripts.stubsabot import fetch_pypi_info, get_github_repo_info


async def run():
    async with aiohttp.ClientSession() as session:
        for p in Path("stubs").iterdir():
            if p.name in ["gdb"]:
                continue
            pypi_info = await fetch_pypi_info(p.name, session)
            repo_info = await get_github_repo_info(session, pypi_info)
            if repo_info:
                replace_metadata(p, repo_info.repo_path)


def replace_metadata(path: Path, repo_path: str) -> None:
    meta_path = path / "METADATA.toml"
    lines = meta_path.read_text().splitlines()
    url = "https://github.com/" + repo_path
    new_lines = [lines[0], f'upstream_repository = "{url}"', *lines[1:]]
    meta_path.write_text("\n".join(new_lines) + "\n")


asyncio.run(run())
```

**Comment by AlexWaygood** [comment]
> Script to add the existing URLs:

Hah, ~~very nice hack~~ excellent reuse of existing code!

> I'll send a followup PR to make stubsabot use the new field.

It would also be great if we could add logic similar to what you've got in your script above to auto-add this new field in `create_baseline_stubs.py`

**Review by AlexWaygood — APPROVED** [review]
Haven't checked every single URL, but can't spot anything wrong from a skim through, and the script you used looks solid!

**Comment by Akuli** [comment]
These files don't have `upstream_repository`:

```
beautifulsoup4/METADATA.toml
braintree/METADATA.toml
cffi/METADATA.toml
docopt/METADATA.toml
docutils/METADATA.toml
editdistance/METADATA.toml
gdb/METADATA.toml
hdbcli/METADATA.toml
humanfriendly/METADATA.toml
ibm-db/METADATA.toml
inifile/METADATA.toml
JACK-Client/METADATA.toml
openpyxl/METADATA.toml
paho-mqtt/METADATA.toml
passlib/METADATA.toml
pexpect/METADATA.toml
protobuf/METADATA.toml
pyaudio/METADATA.toml
pycurl/METADATA.toml
pysftp/METADATA.toml
python-crontab/METADATA.toml
python-nmap/METADATA.toml
pytz/METADATA.toml
tensorflow/METADATA.toml
toposort/METADATA.toml
untangle/METADATA.toml
vobject/METADATA.toml
WebOb/METADATA.toml
```

Should we going to add more URLs in a separate PR or here?

**Comment by srittau** [comment]
Let's do separate PRs. I'd like to merge this now and continue working on stubsabot.

**Pull request merged**